### ISSUE TITLE
Removing and reinstalling the Script Filter dp causes a lot of components to be restarted

### ISSUE THREAD

**Issue by nicolatimeus** [eclipse/kura#1928]
- Component affected:
  - [X] Distribution
    - Installer name kura_3.2.0-SNAPSHOT_raspberry-pi-2-3_installer.deb:
    - Platform: observed on Raspberry pi 3 and emulator
    - Kura version affected Kura 3.2.0 snapshot (probably also previous versions)
    - Java version: java version "1.8.0_65"
Java(TM) SE Runtime Environment (build 1.8.0_65-b17)
Java HotSpot(TM) Client VM (build 25.65-b01, mixed mode)
    - Host OS (e.g. Debian): raspbian
    - Host OS version

  - [ ] Development environment
    - Version: 
    - Operating system (Mac OS, Windows, Linux):
    - OS Version:
    - IDE Version:
    - Host System Java Version:
      
  - [ ] Oomph installer
  - [X] Add-ons
    - Name: Script Filter
    - Version: 1.0.100-SNAPSHOT and 1.0.0
    - URL:
  - [ ] Eclipse Kura Marketplace
  - [ ] Kura Website
  - [ ] Kura Documentation

  - If needed to replicate the issue, please provide references for the browser used:
    - Name:
    - Version:
     
- Textual description of the issue:

Uninstalling and reinstalling the Script Filter dp causes a lot of framework components to be restarted by the Refresh Thread.

```
2018-02-19 14:37:22,860 [Refresh Thread: Equinox Container: 2001f15f-7f15-0018-14db-d15a9290fe56] INFO  o.e.k.c.c.ConfigurationServiceImpl - Registering metatype pid: org.eclipse.kura.wire.CloudPublisher ...
2018-02-19 14:37:22,860 [Refresh Thread: Equinox Container: 2001f15f-7f15-0018-14db-d15a9290fe56] INFO  o.e.k.c.c.ConfigurationServiceImpl - Registering metatype pid: org.eclipse.kura.wire.DbWireRecordStore ...
```

The refresh process takes some time, when it finishes, a new component instance is created for each existing factory component configuration. The new instance will have the same `kura.service.pid` as the pre-esisting instace and a different `service.pid`, this can be observed using the OSGi console:

```
osgi> ls | grep CloudService
12	Active		org.eclipse.kura.cloud.CloudService			org.eclipse.kura.core.cloud(bid=63)
14	Active		org.eclipse.kura.core.cloud.factory.DefaultCloudServiceFactory			org.eclipse.kura.core.cloud(bid=63)
37	Active		org.eclipse.kura.camel.cloud.factory.CamelCloudServiceFactory			org.eclipse.kura.camel.cloud.factory(bid=88)
true
osgi> comp 12 | grep service.pid
      DataService.target = (kura.service.pid=org.eclipse.kura.data.DataService)
      service.pid = org.eclipse.kura.cloud.CloudService-1519050581560-26
      kura.service.pid = org.eclipse.kura.cloud.CloudService
      DataService.target = (kura.service.pid=org.eclipse.kura.data.DataService)
      service.pid = org.eclipse.kura.cloud.CloudService-1519049752820-7
      kura.service.pid = org.eclipse.kura.cloud.CloudService
      DataService.target = (kura.service.pid=org.eclipse.kura.data.DataService)
      service.pid = org.eclipse.kura.cloud.CloudService-1519051038972-45
      kura.service.pid = org.eclipse.kura.cloud.CloudService
true
osgi>
```
The duplication seems to be caused by the fact that the ConfigurationService itself is stopped and restarted by the Refresh Thread:

```
2018-02-19 15:12:18,189 [Refresh Thread: Equinox Container: b08b8dd3-8615-0018-1931-f2ec62533b5f] INFO  o.e.k.c.c.ConfigurationServiceImpl - deactivate...
2018-02-19 15:12:18,190 [Refresh Thread: Equinox Container: b08b8dd3-8615-0018-1931-f2ec62533b5f] INFO  o.e.k.c.c.ConfigurationServiceImpl - Removing component configuration for pid org.eclipse.kura.internal.rest.provider.RestService
2018-02-19 15:12:18,190 [Refresh Thread: Equinox Container: b08b8dd3-8615-0018-1931-f2ec62533b5f] INFO  o.e.k.c.c.ConfigurationServiceImpl - Removing component configuration for pid org.eclipse.kura.broker.artemis.simple.mqtt.BrokerInstance
2018-02-19 15:12:18,191 [Refresh Thread: Equinox Container: b08b8dd3-8615-0018-1931-f2ec62533b5f] INFO  o.e.k.c.c.ConfigurationServiceImpl - Removing component configuration for pid org.eclipse.kura.camel.xml.XmlRouterComponent
2018-02-19 15:12:18,192 [Refresh Thread: Equinox Container: b08b8dd3-8615-0018-1931-f2ec62533b5f] INFO  o.e.k.c.c.ConfigurationServiceImpl - Removing component configuration for pid org.eclipse.kura.broker.artemis.xml.BrokerInstance
2018-02-19 15:12:18,192 [Refresh Thread: Equinox Container: b08b8dd3-8615-0018-1931-f2ec62533b5f] INFO  o.e.k.c.c.ConfigurationServiceImpl - Removing component configuration for pid org.eclipse.kura.watchdog.WatchdogService
2018-02-19 15:12:18,193 [Refresh Thread: Equinox Container: b08b8dd3-8615-0018-1931-f2ec62533b5f] INFO  o.e.k.c.c.ConfigurationServiceImpl - Removing component configuration for pid org.eclipse.kura.db.H2DbService
2018-02-19 15:12:18,194 [Refresh Thread: Equinox Container: b08b8dd3-8615-0018-1931-f2ec62533b5f] INFO  o.e.k.c.c.ConfigurationServiceImpl - Removing component configuration for pid org.eclipse.kura.clock.ClockService
2018-02-19 15:12:18,194 [Refresh Thread: Equinox Container: b08b8dd3-8615-0018-1931-f2ec62533b5f] INFO  o.e.k.c.c.ConfigurationServiceImpl - Removing component configuration for pid org.eclipse.kura.position.PositionService
2018-02-19 15:12:18,511 [Component Resolve Thread (Bundle 7)] INFO  o.e.k.c.c.ConfigurationServiceImpl - activate...
2018-02-19 15:12:18,513 [Component Resolve Thread (Bundle 7)] INFO  o.e.k.c.c.ConfigurationServiceImpl - Loading init configurations from: 1519050602667...
2018-02-19 15:12:18,989 [Component Resolve Thread (Bundle 7)] INFO  o.e.k.c.c.ConfigurationServiceImpl - Creating configuration with pid: conditional and factory pid: org.eclipse.kura.wire.Conditional
2018-02-19 15:12:18,989 [Component Resolve Thread (Bundle 7)] INFO  o.e.k.c.c.ConfigurationServiceImpl - Creating new configuration for factory pid org.eclipse.kura.wire.Conditional and pid conditional
2018-02-19 15:12:18,990 [Component Resolve Thread (Bundle 7)] INFO  o.e.k.c.c.ConfigurationServiceImpl - Updating newly created configuration for pid conditional
2018-02-19 15:12:19,017 [Component Resolve Thread (Bundle 7)] INFO  o.e.k.c.c.ConfigurationServiceImpl - Registering ConfigurableComponent - conditional....
2018-02-19 15:12:19,018 [Component Resolve Thread (Bundle 7)] INFO  o.e.k.c.c.ConfigurationServiceImpl - Registering ConfigurableComponent - conditional....Done
2018-02-19 15:12:19,019 [Component Resolve Thread (Bundle 7)] INFO  o.e.k.c.c.ConfigurationServiceImpl - Creating configuration with pid: timer and factory pid: org.eclipse.kura.wire.Timer
2018-02-19 15:12:19,019 [Component Resolve Thread (Bundle 7)] INFO  o.e.k.c.c.ConfigurationServiceImpl - Creating new configuration for factory pid org.eclipse.kura.wire.Timer and pid timer
2018-02-19 15:12:19,019 [Component Resolve Thread (Bundle 7)] INFO  o.e.k.c.c.ConfigurationServiceImpl - Updating newly created configuration for pid timer
2018
```

The snapshot does not seem to be altered during this process, after a restart Kura behaves normally.

This is probably the same issue reported in #1689.

**Comment by nicolatimeus** [comment]
This issue is likely caused by the localisation, in particular it is caused by the fact that all localisation resources bundles are currently fragments of the `org.eclipse.kura.localization` bundle. 

According to the OSGi spec a bundle/fragment is updated OSGi performs a refresh (stop and restart) of the bundles in its _dependency closure_. The dependency closure contains the list of bundles that directly or indirectly "depend" on the bundles being updated, and is obtainable using the `FrameworkWiring.getDependencyClosure()` method. The dependency closures of the bundles included in the `org.eclipse.kura.script.filter` dp are the following:

```
10:52:22,936 [Component Resolve Thread] INFO  ScriptFilter:90  - script filter dependency closure:
[org.eclipse.kura.wire.script.filter.provider_1.0.100.qualifier [40]]
10:52:22,936 [Component Resolve Thread] INFO  ScriptFilter:92  - localization bundle dependency closure: 
[org.eclipse.kura.wire.devel.component.provider_1.0.4.qualifier [134],
org.eclipse.kura.json.marshaller.unmarshaller.provider_1.0.0.qualifier [116],
org.eclipse.kura.driver.s7plc.provider_1.0.100.qualifier [74],
org.eclipse.kura.asset.helper.provider_1.0.200.qualifier [109],
org.eclipse.kura.localization_1.0.200.qualifier [85],
org.eclipse.kura.core.deployment_1.1.300.qualifier [10],
org.eclipse.kura.driver.gpio.localization_1.0.0.qualifier [119],
org.eclipse.kura.asset.provider_2.0.0.qualifier [45],
org.eclipse.kura.wire.script.filter.provider_1.0.100.qualifier [40],
org.eclipse.kura.wire.helper.provider_1.0.200.qualifier [33],
org.eclipse.kura.driver.gpio.provider_1.0.0.qualifier [56],
org.eclipse.kura.driver.opcua.provider_1.0.200.qualifier [126],
org.eclipse.kura.driver.s7plc.localization_1.0.100.qualifier [100],
org.eclipse.kura.web2_2.0.300.qualifier [76],
org.eclipse.kura.wire.component.join.provider_1.0.0.qualifier [11],
org.eclipse.kura.wire.component.provider_1.0.200.qualifier [37],
org.eclipse.kura.driver.helper.provider_1.0.200.qualifier [78],
org.eclipse.kura.xml.marshaller.unmarshaller.provider_1.0.0.qualifier [34],
org.eclipse.kura.driver.opcua.localization_1.0.200.qualifier [99],
org.eclipse.kura.util_1.0.200.qualifier [17],
org.eclipse.kura.wire.h2db.component.provider_1.0.100.qualifier [124],
org.eclipse.kura.camel.cloud.factory_1.1.200.qualifier [89],
org.eclipse.kura.wire.script.filter.localization_1.0.100.qualifier [29],
org.eclipse.kura.core.configuration_2.0.0.qualifier [97],
org.eclipse.kura.wire.provider_1.0.200.qualifier [5],
org.eclipse.kura.localization.resources_1.0.200.qualifier [42],
org.eclipse.kura.asset.cloudlet.provider_1.0.200.qualifier [130]]
```

The rules for computing the dependency closure are explained in the OSGi spec. Roughly speaking in this case the following should happen:
* Since the `org.eclipse.kura.wire.script.filter.localization` fragment is going to be updated, its host, 
`org.eclipse.kura.localization` belongs to the dependency closure.
* If a bundle A imports a package from bundle B that belongs to the dependency closure, then A belongs to the dependency closure as well.

`org.eclipse.kura.core.configuration` is in the closure because it imports the  `org.eclipse.kura.core.util` package from the `org.eclipse.kura.util` bundle, which in turn imports the `org.eclipse.kura.localization`package.

This should also explain why the Web Ui is restarted in scenario shown in issue #1689.

**Comment by pierantoniomerlino** [comment]
@nicolatimeus Great explanation! It also happens with different types of bundles (i.e. SensorTag Driver).

**Comment by nicolatimeus** [comment]
A short term solution to this issue can be removing localisation from core components in order to prevent them from being restarted.

**Comment by cdealti** [comment]
@nicolatimeus can you please start removing the dependencies from org.eclipse.kura.localization so that the Configuration Service and the Web UI are not deactivated?
This should provide a temporary fix for 3.2.0.
